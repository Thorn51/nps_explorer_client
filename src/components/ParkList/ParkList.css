/* Colors */
:root {
  --darkGunmetal: #122932;
  --blackCoral: #576066;
  --queenBlue: #386fa4;
  --middleBlue: #80ced7;
  --mintCream: #f7fff7;
  --platinum: #e0e2e3;
}

.parks_container {
  width: 90%;
  margin: 0 auto 0 auto;
}

.parks_list_header {
  text-align: center;
  padding: 15px;
}
