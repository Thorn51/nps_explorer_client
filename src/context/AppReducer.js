export default (state, action) => {
  switch (action.type) {
    case "SEARCH_STATE":
      return {
        ...state,
        searchState: action.payload,
        loading: true
      };
    case "PARKS_IN_STATE":
      return {
        ...state,
        parksInState: action.payload,
        loading: false
      };
    case "NPS_NEWS":
      return {
        ...state,
        news: [action.payload],
        loadingNews: false
      };
    case "PARK":
      return {
        ...state,
        park: action.payload[0],
        loadingPark: false
      };
    case "ADD_FAVORITE":
      return {
        ...state,
        favorite: [{ ...action.payload }]
      };
    case "FETCH_PARKS_ERROR":
      return {
        ...state,
        error: action.payload
      };
    case "FETCH_NEWS_ERROR":
      return {
        ...state,
        error: action.payload
      };
    case "FETCH_PARK_ERROR":
      return {
        ...state,
        error: action.payload
      };
    case "POST_COMMENT":
      return {
        ...state,
        comments: [...state.comments, action.payload]
      };
    case "LOGIN_ERROR":
      return {
        ...state,
        error: action.payload
      };
    default:
      return state;
  }
};
