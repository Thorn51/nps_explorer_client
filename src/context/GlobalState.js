import React, { createContext, useReducer } from "react";
import AppReducer from "./AppReducer";
import npsApiService from "../services/npsApiServices";

// Initial state
const initialState = {
  searchState: [],
  parksInState: [],
  news: [],
  park: [],
  favorite: [],
  comments: [],
  error: null,
  loading: null,
  loadingNews: true,
  loadingPark: true
};

// Create context
export const GlobalContext = createContext(initialState);

// Provider component
export const GlobalProvider = ({ children }) => {
  const [state, dispatch] = useReducer(AppReducer, initialState);

  // Action -> Select state and store it
  function selectState(stateCode, stateName) {
    dispatch({
      type: "SEARCH_STATE",
      payload: { stateCode, stateName }
    });
  }

  // Action -> Fetch the parks by state code
  async function getParks(stateCode) {
    try {
      let parks = await npsApiService.getParksByState(stateCode);

      dispatch({
        type: "PARKS_IN_STATE",
        payload: parks.data
      });
    } catch (error) {
      console.log(error);
      dispatch({
        type: "FETCH_PARKS_ERROR",
        payload: error.response.data
      });
    }
  }

  // Action -> Fetch news
  async function getNews() {
    try {
      let news = await npsApiService.getNews();
      dispatch({
        type: "NPS_NEWS",
        payload: news.data
      });
    } catch (error) {
      console.log(error);
      dispatch({
        type: "FETCH_NEWS_ERROR",
        payload: error.response.data
      });
    }
  }

  // Action -> fetch park by park code for park page
  async function getParkByParkCode(parkCode) {
    try {
      let park = await npsApiService.getParkByCode(parkCode);
      dispatch({
        type: "PARK",
        payload: park.data
      });
    } catch (error) {
      dispatch({
        type: "PARK_FETCH_ERROR",
        payload: error
      });
    }
  }

  // Action -> add park to favorites
  function addFavorite(favorite) {
    dispatch({
      type: "ADD_FAVORITE",
      payload: favorite
    });
  }

  // Action -> post user comment
  function postComment(newComment) {
    dispatch({
      type: "POST_COMMENT",
      payload: newComment
    });
  }

  return (
    <GlobalContext.Provider
      value={{
        searchState: state.searchState,
        parksInState: state.parksInState,
        news: state.news,
        park: state.park,
        error: state.error,
        loading: state.loading,
        loadingNews: state.loadingNews,
        loadingPark: state.loadingPark,
        comments: state.comments,
        favorite: state.favorite,
        selectState,
        getParks,
        getNews,
        getParkByParkCode,
        postComment,
        addFavorite
      }}
    >
      {children}
    </GlobalContext.Provider>
  );
};
