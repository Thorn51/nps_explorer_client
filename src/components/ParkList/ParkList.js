import React, { useContext } from "react";
import { GlobalContext } from "../../context/GlobalState";
import ParkCard from "../ParkCard/ParkCard";
import "./ParkList.css";

export default function ParkList() {
  const { parksInState, searchState } = useContext(GlobalContext);

  return (
    <div className="parks_container">
      <h3 className="parks_list_header">
        {searchState[0].stateName} ({parksInState[0].length})
      </h3>
      <section className="featured_parks">
        {parksInState[0].map(park => (
          <ParkCard
            key={park.id}
            designation={park.designation}
            name={park.name}
            description={park.description}
            parkCode={park.parkCode}
          />
        ))}
      </section>
    </div>
  );
}
