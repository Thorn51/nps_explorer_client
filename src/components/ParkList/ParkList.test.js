import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import ParkList from "./ParkList";

import { GlobalProvider } from "../../context/GlobalState";

describe.skip("<ParkList />", () => {
  it("Renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <GlobalProvider>
        <ParkList />
      </GlobalProvider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("Renders the UI as expected", () => {
    const tree = renderer.create(<ParkList />).toJSON();
    expect("tree").toMatchSnapshot();
  });
});
