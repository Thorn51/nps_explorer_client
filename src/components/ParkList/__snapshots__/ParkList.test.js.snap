// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParkList /> Renders the UI as expected 1`] = `"tree"`;
